kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/AdapterGenerator.kt

```diff
@@ -159,6 +160,19 @@ internal class AdapterGenerator(
         .build()
   }
 
+  private fun generateJsonDataExceptionFun(): FunSpec {
```

Comment:
Can we do this without the extra generated function? I'm open to the idea of this but adding a copy of this method in every class probably adds more bloat than it saves. Either extracting this to something like Util, or I bet we could get by with kotlin's standard string APIs directly

Reply:
The method is likely cheaper size-wise than duplicating. For performance we don't care. For method count it's unfortunate. Not sure if R8 will outline to a single method.

Reply:
what if we make it `inline`?

Reply:
Inline is only useful for library functions. We already have inline available in the code generator by either generating it twice or once in a method. If you generate it twice, R8 will probably outline it as I believe it special-cases 1-6 argument `StringBuilder` usage.

Reply:
https://r8.googlesource.com/r8/+/master/src/main/java/com/android/tools/r8/utils/InternalOptions.java#553

Code must be between 3 and 99 bytes and occur 20 times. Probably best to just generate it multiple times then to trigger outlining sooner.

Reply:
I think it would be possible to move it to the Util class in Moshi. I am not sure how outlining works, is it only outlining within one class or would it also deduplicate this method in multiple adapter classes? Maybe if it were static?

Reply:
Outlining works on code patterns, not methods, so it would find the StringBuilder and 6 append statements duplicated 2*adapter count times, pull those bytecodes out into a single static method, and rewrite all adapters to use it.